pkgs/build-support/gradle/update-deps.nix

```
@@ -0,0 +1,147 @@
+  nixShellKeep = lib.concatMapStringsSep " " (x: "--keep ${x}") keep;
+in
+writeShellScript "fetch-deps.sh" ''
+  set -eo pipefail
+  ${lib.optionalString (lib.hasInfix "$attrPath" code) ''
+    attrPath=${lib.escapeShellArg attrPath}
+    [[ -n "$attrPath" ]] || attrPath="$1"
+    [[ -n "$attrPath" ]] || attrPath="$UPDATE_NIX_ATTR_PATH"
+    [[ -n "$attrPath" ]] || (echo "Set UPDATE_NIX_ATTR_PATH or pass attr path as the first argument!" && exit 1)
+  ''}
+  outPath="${
+    # if this is an absolute path in nix store, use path relative to the output root
+    if lib.hasPrefix "${builtins.storeDir}/" (toString depsPath)
+    then builtins.concatStringsSep "/" (lib.drop 1 (lib.splitString "/" (lib.removePrefix "${builtins.storeDir}/" (toString depsPath))))
+    # if this is an absolute path, just use that path
+    else if lib.hasPrefix "/" (toString depsPath)
+    then toString depsPath
+    # otherwise, use a path relative to the package
+    else ''$(nix-instantiate --eval -E "dirOf (${code}).meta.position" | tr -d '"')/${depsPath}''
+  }"
+
+  nix-build --no-out-link -E "${code'}" > /dev/null
+
+  pushd "$(${coreutils}/bin/mktemp -d)"
+  MITM_CACHE_DIR="$PWD"
+  trap "${coreutils}/bin/rm -rf '$MITM_CACHE_DIR'" SIGINT SIGTERM ERR EXIT
+  ${openssl}/bin/openssl genrsa -out ca.key 2048
+  ${openssl}/bin/openssl req -x509 -new -nodes -key ca.key -sha256 -days 1 -out ca.cer -subj "/C=AL/ST=a/L=a/O=a/OU=a/CN=example.org"
+  export MITM_CACHE_HOST=127.0.0.1
+  export MITM_CACHE_PORT="''${mitmCachePort:-$(${python3.pkgs.ephemeral-port-reserve}/bin/ephemeral-port-reserve "$MITM_CACHE_HOST")}"
+  export MITM_CACHE_ADDRESS="$MITM_CACHE_HOST:$MITM_CACHE_PORT"
+  # forget all redirects - this makes the lockfiles predictable
+  # not only does this strip CDN URLs, but it also improves security - since the redirects aren't
+  # stored in the lockfile, a malicious actor can't change the redirect URL stored in the lockfile
+  ${mitm-cache}/bin/mitm-cache \
+    -l"$MITM_CACHE_ADDRESS" \
+    record \
+    --reject '\.(md5|sha(1|256|512:?):?)$' \
+    --forget-redirects-from '.*' \
+    --record-text '/maven-metadata\.xml$' >/dev/null 2>/dev/null &
+  MITM_CACHE_PID="$!"
+  # wait for mitm-cache to fully start
+  for i in {0..20}; do
+    ${procps}/bin/ps -p "$MITM_CACHE_PID" >/dev/null || (echo "Failed to start mitm-cache" && exit 1)
+    ${curl}/bin/curl "$MITM_CACHE_ADDRESS" && break
+    [[ "$i" -eq 20 ]] && (echo "Failed to start mitm-cache" && exit 1)
+    sleep 0.5
+  done
+  trap "kill '$MITM_CACHE_PID'" SIGINT SIGTERM ERR EXIT
+  export MITM_CACHE_CERT_DIR="$PWD"
+  export MITM_CACHE_CA="$MITM_CACHE_CERT_DIR/ca.cer"
+  popd
+  useBwrap="''${USE_BWRAP:-${toString useBwrap}}"
+  nixShell="$(${coreutils}/bin/mktemp --suffix=.nix)"
+  trap "${coreutils}/bin/rm -f '$nixShell'" SIGINT SIGTERM ERR EXIT
+  echo "${code'}" > "$nixShell"
```

You could evaluate first, outside the sandbox.

```suggestion
  echo "${code'}" > "$nixShell"
  nixShellDrv=$(nix-instantiate "$nixShell")
```

That should do the trick, but you could simplify it further by performing the second evaluation immediately:

```suggestion
  nixShellDrv="${builtins.unsafeDiscardOutputDependency pkg.drvPath}"
```

`pkg` would be a new parameter that replaces `code` and `attrPath`, though you could keep them around for a while to make the out-of-tree migration easier.

---

Why is `builtins.unsafeDiscardOutputDependency` used here?

---

Thanks, this works, although `pkg` can be clunky to pass to `updateScript`, so I left the `pname` argument